src/dotnet-interactive-vscode/common/vscode/extension.ts

```diff
@@ -142,7 +142,7 @@ export async function activate(context: vscode.ExtensionContext) {
     context.subscriptions.push(vscode.workspace.onDidRenameFiles(e => handleFileRenames(e, clientMapper)));
 
     // language registration
-    context.subscriptions.push(vscode.workspace.onDidOpenTextDocument(async e => await updateNotebookCellLanguageInMetadata(e)));
```

Comment:
Thank you @jrieken!
The ` vscode.workspace.onDidOpenTextDocument` causes the dirty and unsaved notebook to be reopen forever. The suggestion to use the `vscode.notebook.onDidOpenNotebookDocument` event also let us batch the edits as a single apply call